### PR TITLE
Add support for state_channel txns

### PR DESCRIPTION
**Pull request by vihu**
- Add migrations for state_channel txns
- Handle two new txn types `state_channel_open` and `state_channel_close`
- Add txn actor roles for the two new txn types

### CODE REVIEW

File: src/be_txn.erl [hunk 1]
```diff
@@ -214,4 +218,33 @@ to_json(blockchain_txn_payment_v2, T, _Ledger) ->
...
+                             <<"num_bytes">> => NumBytes
+                           }
+                  end,
+
+    SCJson = fun(SC) ->
+                     #{<<"id">> => ?BIN_TO_B64(blockchain_state_channel_v1:id(SC)),
+                       <<"owner">> => ?BIN_TO_B58(blockchain_state_channel_v1:owner(SC)),
+                       <<"credits">> => blockchain_state_channel_v1:credits(SC),
+                       <<"balances">> => [BalanceJson(B) || B <- blockchain_state_channel_v1:balances(SC)],
+                       <<"root_hash">> => blockchain_state_channel_v1:root_hash(SC),
```
Comment (madninja) [inline src/be_txn.erl]: B64 for root hash?

Comment (vihu) [reply src/be_txn.erl]: yup, updated :+1:

File: src/be_txn.erl [hunk 2]
```diff
@@ -214,4 +218,33 @@ to_json(blockchain_txn_payment_v2, T, _Ledger) ->
...
+    #{<<"id">> => ?BIN_TO_B64(blockchain_txn_state_channel_open_v1:id(T)),
+      <<"owner">> => ?BIN_TO_B58(blockchain_txn_state_channel_open_v1:owner(T)),
+      <<"amount">> => blockchain_txn_state_channel_open_v1:amount(T),
+      <<"fee" >> => blockchain_txn_state_channel_open_v1:fee(T),
+      <<"nonce">> => blockchain_txn_state_channel_open_v1:nonce(T),
+      <<"expire_within">> => blockchain_txn_state_channel_open_v1:expire_within(T),
+      <<"signature">> => ?BIN_TO_B64(blockchain_txn_state_channel_open_v1:signature(T)) };
+to_json(blockchain_txn_state_channel_close_v1, T, _Ledger) ->
+
+    BalanceJson = fun(Balance) ->
```
Comment (vihu) [inline src/be_txn.erl:232]: This is stale now and needs to be "summaries" or something

File: rebar.config [hunk 3]
```diff
@@ -25,7 +25,7 @@
   {telemetry, "0.4.1"},
   {psql_migration, {git, "https://github.com/helium/psql-migration.git", {branch, "master"}}},
   {clique, {git, "https://github.com/helium/clique.git", {branch, "develop"}}},
-  {blockchain, {git, "https://github.com/helium/blockchain-core.git", {branch, "master"}}},
+  {blockchain, {git, "https://github.com/helium/blockchain-core.git", {branch, "rg/sc-debug"}}},
```
Comment (vihu) [inline rebar.config:28]: TODO: Switch to `rg/data-credits` core branch

File: src/be_txn.erl [hunk 4]
```diff
@@ -57,10 +61,11 @@ to_json(blockchain_txn_security_coinbase_v1, T, _Ledger) ->
 to_json(blockchain_txn_oui_v1, T, _Ledger) ->
     #{<<"owner">> => ?BIN_TO_B58(blockchain_txn_oui_v1:owner(T)),
       <<"addresses">> => [binary_to_list(Addr) || Addr <- blockchain_txn_oui_v1:addresses(T)],
```
Comment (madninja) [inline src/be_txn.erl:63]: What kind of addresses are these? Are they router_addresses? If so, BIN_TO_B58?

File: src/be_txn.erl [hunk 5]
```diff
@@ -69,10 +74,29 @@ to_json(blockchain_txn_gen_gateway_v1, T, _Ledger) ->
       <<"location">> => ?MAYBE_H3(blockchain_txn_gen_gateway_v1:location(T)),
       <<"nonce">> => blockchain_txn_gen_gateway_v1:nonce(T) };
 to_json(blockchain_txn_routing_v1, T, _Ledger) ->
+
+    ActionJson = fun(Action) ->
+                         case Action of
+                             {update_routers, RouterAddresses} ->
+                                 #{<<"action">> => <<"update_routers">>,
+                                   <<"router_addresses">> =>[?BIN_TO_B58(A) || A <- RouterAddresses]};
```
Comment (madninja) [inline src/be_txn.erl:82]: How does this related to "addresses" in the oui txn above? Consolidate?

File: src/be_txn.erl [hunk 6]
```diff
@@ -214,4 +238,34 @@ to_json(blockchain_txn_payment_v2, T, _Ledger) ->
...
+      <<"owner">> => ?BIN_TO_B58(blockchain_txn_state_channel_open_v1:owner(T)),
+      <<"oui">> => blockchain_txn_state_channel_open_v1:oui(T),
+      <<"fee" >> => blockchain_txn_state_channel_open_v1:fee(T),
+      <<"nonce">> => blockchain_txn_state_channel_open_v1:nonce(T),
+      <<"expire_within">> => blockchain_txn_state_channel_open_v1:expire_within(T),
+      <<"signature">> => ?BIN_TO_B64(blockchain_txn_state_channel_open_v1:signature(T)) };
+to_json(blockchain_txn_state_channel_close_v1, T, _Ledger) ->
+
+    SummaryJson = fun(Summary) ->
+                          #{ <<"client_pubkeybin">> => ?BIN_TO_B58(blockchain_state_channel_summary_v1:client_pubkeybin(Summary)),
```
Comment (madninja) [inline src/be_txn.erl:253]: Don't call it pubkey_bin.. Nothing else has type info in it like this.. it's the `client`, right? or is there something specific that we need to know here

File: migrations/1583860178-state_channel.sql [hunk 7]
```diff
@@ -0,0 +1,9 @@
+-- migrations/1583860178-state_channel.sql
+-- :up
+-- Up migration
+ALTER TYPE transaction_type ADD VALUE 'state_channel_open_v1';
```
Comment (madninja) [inline migrations/1583860178-state_channel.sql:4]: Rename files to newer timestamps

File: rebar.config [hunk 8]
```diff
@@ -25,7 +25,7 @@
   {telemetry, "0.4.1"},
   {psql_migration, {git, "https://github.com/helium/psql-migration.git", {branch, "master"}}},
   {clique, {git, "https://github.com/helium/clique.git", {branch, "develop"}}},
-  {blockchain, {git, "https://github.com/helium/blockchain-core.git", {branch, "master"}}},
+  {blockchain, {git, "https://github.com/helium/blockchain-core.git", {branch, "rg/data-credits"}}},
```
Comment (madninja) [inline rebar.config:28]: Fix this before merge

File: src/be_txn.erl [hunk 9]
```diff
@@ -43,7 +43,11 @@ to_type(blockchain_txn_dc_coinbase_v1) ->
 to_type(blockchain_txn_token_burn_exchange_rate_v1) ->
     "token_burn_exchange_rate_v1";
```
Comment (madninja) [inline src/be_txn.erl:44]: Is this still valid?

File: src/be_db_txn_actor.erl [hunk 10]
```diff
@@ -133,4 +133,11 @@ to_actors(blockchain_txn_payment_v2, T) ->
...
     lists:foldl(ToActors, [{"payer", blockchain_txn_payment_v2:payer(T)}],
-                blockchain_txn_payment_v2:payments(T)).
+                blockchain_txn_payment_v2:payments(T));
+to_actors(blockchain_txn_state_channel_open_v1, T) ->
+    %% NOTE: owner => sc_opener, renamed for better reasoning
+    [{"sc_opener", blockchain_txn_state_channel_open_v1:owner(T)} ];
+to_actors(blockchain_txn_state_channel_close_v1, T) ->
+    %% NOTE: closer can be one of the clients of the state channel or the owner of the router
+    %% if the state_channel expires
+    [{"sc_closer", blockchain_txn_state_channel_close_v1:closer(T)} ].
```
Comment (madninja) [inline src/be_db_txn_actor.erl:143]: This is missing a number of actors for the token burn?

File: src/be_txn.erl (hunk 5 above)
Comment (vihu) [reply src/be_txn.erl:82]: Sorry but consolidate what exactly? For the `routing_txn` there are multiple actions which can be performed, I couldn't think of a way to represent it better in json form. For more context, https://github.com/helium/blockchain-core/pull/434/files#diff-79d16094f5665a5f02f76609f17b0b9cR37 that's the thing we're trying to expose here

File: src/be_txn.erl (hunk 9 above)
Comment (vihu) [reply src/be_txn.erl:44]: Well not quite but we have it in the core anyway, hence I put it here. We can get back to it when DCs gain value.

File: src/be_db_txn_actor.erl (hunk 10 above)
Comment (vihu) [reply src/be_db_txn_actor.erl:143]: I don't understand, actors for `blockchain_txn_token_burn_v1` are already defined in previous clauses.

File: src/be_txn.erl (hunk 4 above)
Comment (vihu) [reply src/be_txn.erl]: Added in 8bec535

File: migrations/1583860178-state_channel.sql (hunk 7 above)
Comment (vihu) [reply migrations/1583860178-state_channel.sql]: Added in 8bec535

File: src/be_txn.erl (hunk 6 above)
Comment (vihu) [reply src/be_txn.erl]: Renamed 8bec535

File: rebar.config (hunk 8 above)
Comment (vihu) [reply rebar.config:28]: Yeah, once core branch lands will do it :+1:

File: src/be_txn.erl (hunk 5 above)
Comment (madninja) [reply src/be_txn.erl:82]: Sorry, I meant to use the same key for both `router_addresses` and the `addresses` in the oui txn above. It's minor

File: src/be_db_txn_actor.erl (hunk 10 above)
Comment (madninja) [reply src/be_db_txn_actor.erl:143]: I was confusing clauses, ignore

File: src/be_txn.erl [hunk 11]
```diff
@@ -214,4 +238,34 @@ to_json(blockchain_txn_payment_v2, T, _Ledger) ->
...
+                           }
+                  end,
+
+    SCJson = fun(SC) ->
+                     #{<<"id">> => ?BIN_TO_B64(blockchain_state_channel_v1:id(SC)),
+                       <<"owner">> => ?BIN_TO_B58(blockchain_state_channel_v1:owner(SC)),
+                       <<"nonce">> => blockchain_state_channel_v1:nonce(SC),
+                       <<"summaries">> => [SummaryJson(B) || B <- blockchain_state_channel_v1:summaries(SC)],
+                       <<"root_hash">> => ?BIN_TO_B64(blockchain_state_channel_v1:root_hash(SC)),
+                       <<"state">> => blockchain_state_channel_v1:state(SC),
```
Comment (madninja) [inline src/be_txn.erl:265]: Can you actually have strings for the states here? like `<<"open">>` and `<<''closed">>`? It'll make queries easier

**Pull request merged**